// @gnosis.pm/Rsafe-core-sdk/dist/src/managers/contractManager.js
"use strict";
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _ContractManager_contractNetworks, _ContractManager_isL1SafeMasterCopy, _ContractManager_safeContract, _ContractManager_multiSendContract;
Object.defineProperty(exports, "__esModule", { value: true });
const config_1 = require("../contracts/config");
const safeDeploymentContracts_1 = require("../contracts/safeDeploymentContracts");
class ContractManager {
    constructor() {
        _ContractManager_contractNetworks.set(this, void 0);
        _ContractManager_isL1SafeMasterCopy.set(this, void 0);
        _ContractManager_safeContract.set(this, void 0);
        _ContractManager_multiSendContract.set(this, void 0);
    }
    static async create({ ethAdapter, safeAddress, isL1SafeMasterCopy, contractNetworks }) {
        const contractManager = new ContractManager();
        await contractManager.init({ ethAdapter, safeAddress, isL1SafeMasterCopy, contractNetworks });
        return contractManager;
    }
    async init({ ethAdapter, safeAddress, isL1SafeMasterCopy, contractNetworks }) {
        const chainId = await ethAdapter.getChainId();
        const customContracts = contractNetworks === null || contractNetworks === void 0 ? void 0 : contractNetworks[chainId];
        __classPrivateFieldSet(this, _ContractManager_contractNetworks, contractNetworks, "f");
        __classPrivateFieldSet(this, _ContractManager_isL1SafeMasterCopy, isL1SafeMasterCopy, "f");
        const temporarySafeContract = await this.getSafeContract({
            ethAdapter,
            safeVersion: config_1.SAFE_LAST_VERSION,
            chainId,
            safeAddress,
            isL1SafeMasterCopy,
            customContracts
        });
        const safeVersion = await temporarySafeContract.getVersion();
        __classPrivateFieldSet(this, _ContractManager_safeContract, await this.getSafeContract({
            ethAdapter,
            safeVersion,
            chainId,
            safeAddress,
            isL1SafeMasterCopy,
            customContracts
        }), "f");
        __classPrivateFieldSet(this, _ContractManager_multiSendContract, await this.getMultiSendContract({
            ethAdapter,
            safeVersion,
            chainId,
            customContracts
        }), "f");
    }
    get contractNetworks() {
        return __classPrivateFieldGet(this, _ContractManager_contractNetworks, "f");
    }
    get isL1SafeMasterCopy() {
        return __classPrivateFieldGet(this, _ContractManager_isL1SafeMasterCopy, "f");
    }
    get safeContract() {
        return __classPrivateFieldGet(this, _ContractManager_safeContract, "f");
    }
    get multiSendContract() {
        return __classPrivateFieldGet(this, _ContractManager_multiSendContract, "f");
    }
    async getSafeContract({ ethAdapter, safeVersion, chainId, safeAddress, isL1SafeMasterCopy, customContracts }) {
        const safeSingletonDeployment = (0, safeDeploymentContracts_1.getSafeContractDeployment)(safeVersion, chainId, isL1SafeMasterCopy);
        const temporarySafeContract = ethAdapter.getSafeContract({
            safeVersion: config_1.SAFE_LAST_VERSION,
            chainId,
            singletonDeployment: safeSingletonDeployment,
            customContractAddress: safeAddress,
            customContractAbi: customContracts === null || customContracts === void 0 ? void 0 : customContracts.safeMasterCopyAbi
        });
        const isContractDeployed = await ethAdapter.isContractDeployed(temporarySafeContract.getAddress());
        if (!isContractDeployed) {
            throw new Error('Safe Proxy contract is not deployed in the current network');
        }
        return temporarySafeContract;
    }
    async getMultiSendContract({ ethAdapter, safeVersion, chainId, customContracts }) {
        const multiSendDeployment = (0, safeDeploymentContracts_1.getMultiSendContractDeployment)(safeVersion, chainId);
        const multiSendContract = await ethAdapter.getMultiSendContract({
            safeVersion,
            chainId,
            singletonDeployment: multiSendDeployment,
            customContractAddress: customContracts === null || customContracts === void 0 ? void 0 : customContracts.multiSendAddress,
            customContractAbi: customContracts === null || customContracts === void 0 ? void 0 : customContracts.multiSendAbi
        });
        const isContractDeployed = await ethAdapter.isContractDeployed(multiSendContract.getAddress());
        if (!isContractDeployed) {
            throw new Error('Multi Send contract is not deployed in the current network');
        }
        return multiSendContract;
    }
}
_ContractManager_contractNetworks = new WeakMap(), _ContractManager_isL1SafeMasterCopy = new WeakMap(), _ContractManager_safeContract = new WeakMap(), _ContractManager_multiSendContract = new WeakMap();
exports.default = ContractManager;
